draft-irtf-cfrg-hpke.md

```diff
@@ -271,52 +273,108 @@ A set of algorithm identifiers for concrete instantiations of these
+  labeled_IKM = concat("RFCXXXX ", label, IKM)
+  return Extract(salt, labeled_IKM)
+
+def LabeledExpand(PRK, label, info, L):
+  labeled_info = concat(encode_big_endian(L, 2),
+                        "RFCXXXX ", label, info)
+  return Expand(PRK, labeled_info, L)
+~~~
+\[\[RFC editor: please change "RFCXXXX" to the correct number before publication.]]
+
 ## DH-Based KEM
 
-Suppose we are given a Diffie-Hellman group that provides the
+Suppose we are given a KDF, and a Diffie-Hellman group providing the
 following operations:
 
 - GenerateKeyPair(): Generate an ephemeral key pair `(sk, pk)`
   for the DH group in use
 - DH(sk, pk): Perform a non-interactive DH exchange using the
-  private key sk and public key pk to produce a shared secret
-  of length Npk
-- Marshal(pk): Produce a fixed-length byte string
+  private key sk and public key pk to produce a Diffie-Hellman
+  shared secret of length Ndh
+- Marshal(pk): Produce a byte string of length Npk
   encoding the public key `pk`
-- Unmarshal(enc): Parse a fixed-length byte string to recover a
+- Unmarshal(enc): Parse a byte string of length Npk to recover a
   public key
+- Ndh: The length in bytes of a Diffie-Hellman shared secret produced
+  by the DH function of this KEM.
 
-Then we can construct a KEM (which we'll call "DHKEM") in the
-following way:
+Then we can construct a KEM called `DHKEM(Group, KDF)` in the
+following way, where `Group` denotes the Diffie-Hellman group and
+`KDF` the KDF:
 
 ~~~
+def ExtractAndExpand(dh, context_kem):
+  prk = LabeledExtract(0, "dh", dh)
+  return LabeledExpand(prk, "prk", context_kem, Nzz)
+
 def Encap(pkR):
   skE, pkE = GenerateKeyPair()
-  zz = DH(skE, pkR)
+  dh = DH(skE, pkR)
   enc = Marshal(pkE)
+
+  pkRm = Marshal(pkR)
+  context_kem = concat(enc, pkRm)
+
+  zz = ExtractAndExpand(dh, context_kem)
   return zz, enc
 
-def Decap(enc, skR):
+def Decap(enc, skR, pkR):
   pkE = Unmarshal(enc)
-  return DH(skR, pkE)
+  dh = DH(skR, pkE)
 
-def AuthEncap(pkR, skS):
+  pkRm = Marshal(pkR)
+  context_kem = concat(enc, pkRm)
+
+  zz = ExtractAndExpand(dh, context_kem)
+  return zz, enc
+
+def AuthEncap(pkR, skS, pkS):
   skE, pkE = GenerateKeyPair()
-  zz = concat(DH(skE, pkR), DH(skS, pkR))
+  dh = concat(DH(skE, pkR), DH(skS, pkR))
   enc = Marshal(pkE)
+
+  pkRm = Marshal(pkR)
+  pkSm = Marshal(pkS)
+  context_kem = concat(enc, pkRm, pkSm)
+
+  zz = ExtractAndExpand(dh, context_kem)
   return zz, enc
 
-def AuthDecap(enc, skR, pkS):
+def AuthDecap(enc, skR, pkR, pkS):
   pkE = Unmarshal(enc)
-  return concat(DH(skR, pkE), DH(skR, pkS))
+  dh = concat(DH(skR, pkE), DH(skR, pkS))
+
+  pkRm = Marshal(pkR)
+  pkSm = Marshal(pkS)
+  context_kem = concat(enc, pkRm, pkSm)
+
+  zz = ExtractAndExpand(dh, context_kem)
+  return zz, enc
 ~~~
 
+The KDF used in DHKEM can be equal to or different from the KDF used
+in the remainder of HPKE depending on the chosen variant. See
+{{domain-separation}} for a security-related discussion.
```

Comment:
Do we have guidance on what the right answer is?

Reply:
In this pull request, I suggest one particular KDF for each of DHKEM's DH groups. This recommendation is based on the security level of the DH group. We could include as guidance that choosing the same KDF for the remainder of KDF is the best trade-of: this keeps the security level of DHKEM while not pretending a higher security level by using a KDF with greater output length.

Does this go into the direction you meant to go with your question?